### PR TITLE
Linter suggestions

### PR DESCRIPTION
**Pull request by manuelpuyol**
Related to https://github.com/primer/view_components/issues/623

### Summary

This PR starts to work on linter suggestions, providing a better feedback for users on how to use our components.

| Before | After |
|--------|--------|
| image | image  |

### When does it work?

For now, the linter suggestion supports:

1. classes that the component generate (not System Arguments / custom classes)
2. `aria-` attributes
3. `data-` attributes
4. `test_selector` ERB block
5. `tag` attribute
6. `disabled` and `type` (for Button only)

### TODOs

1. I'm not looking at the content, so the "suggestion" is not auto-correctable since I still don't know how to look at the whole element block (any AST experts here? 😄 ).
2. Integrating with https://github.com/primer/view_components/pull/612
3. Dealing with slots

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,10 @@
 
 ### Misc
 
+* Add linter suggestions for `Button` component.
```
Comment (joelhawksley) [inline CHANGELOG.md:15]: This seems like a bigger deal than `misc` 😄

File: lib/primer/view_components/linters/argument_mappers/button.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+require_relative "conversion_error"
+require_relative "system_arguments"
+
+module ERBLint
+  module Linters
+    module ArgumentMappers
+      # Maps classes in a button element to arguments for the Button component.
+      class Button
+        SCHEME_MAPPINGS = {
```
Comment (joelhawksley) [inline lib/primer/view_components/linters/argument_mappers/button.rb:11]: Do you think we could reuse the scheme mappings from the ButtonComponent? (I can also see why we'd rather not include the component for dependency reasons)

My concern would be about keeping these in sync.

File: lib/primer/view_components/linters/argument_mappers/button.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            elsif class_name == "btn-block"
+              acc[:block] = true
+            elsif class_name == "BtnGroup-item"
+              acc[:group_item] = true
+            else
+              raise ConversionError, "Cannot convert class \"#{class_name}\""
+            end
+          end
+
+          hash
```
Comment (joelhawksley) [inline lib/primer/view_components/linters/argument_mappers/button.rb:79]: Does this work?

```suggestion
          classes.value.split(" ").each_with_object({}) do |class_name, acc|
            next if class_name == "btn"

            if SCHEME_MAPPINGS[class_name] && acc[:scheme].nil?
              acc[:scheme] = SCHEME_MAPPINGS[class_name]
            elsif VARIANT_MAPPINGS[class_name] && acc[:variant].nil?
              acc[:variant] = VARIANT_MAPPINGS[class_name]
            elsif class_name == "btn-block"
              acc[:block] = true
            elsif class_name == "BtnGroup-item"
              acc[:group_item] = true
            else
              raise ConversionError, "Cannot convert class \"#{class_name}\""
            end
          end
```

File: lib/primer/view_components/linters/argument_mappers/conversion_error.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+module ERBLint
+  module Linters
+    module ArgumentMappers
+      # Error when converting atguments.
```
Comment (joelhawksley) [inline lib/primer/view_components/linters/argument_mappers/conversion_error.rb:6]:
```suggestion
      # Error when converting arguments.
```

File: lib/primer/view_components/linters/argument_mappers/system_arguments.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        TEST_SELECTOR_REGEX = /test_selector\((?<selector>.+)\)$/.freeze
+
+        attr_reader :attribute
+        def initialize(attribute)
+          @attribute = attribute
+        end
+
+        def to_args
+          if attribute.erb?
+            _, _, code_node, = *attribute.node
```
Comment (joelhawksley) [inline lib/primer/view_components/linters/argument_mappers/system_arguments.rb:20]:
```suggestion
            _, _, code_node = *attribute.node
```

File: lib/primer/view_components/linters/argument_mappers/system_arguments.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require_relative "conversion_error"
+
+module ERBLint
+  module Linters
+    module ArgumentMappers
+      # Maps classes in a button element to arguments for the Button component.
```
Comment (joelhawksley) [inline lib/primer/view_components/linters/argument_mappers/system_arguments.rb:8]: Is this the right description?

File: lib/primer/view_components/linters/argument_mappers/button.rb (hunk 2 above)
Comment (manuelpuyol) [reply lib/primer/view_components/linters/argument_mappers/button.rb:11]: the problem is everything that we need to load to get access to those constants :/

**Review by joelhawksley — APPROVED** [review]
👏🏻 fantastic!

Do you think it's worth waiting on the system arguments mapping work from @jonrohan before we roll this out?

**Comment by manuelpuyol** [comment]
> 👏🏻 fantastic!
> 
> Do you think it's worth waiting on the system arguments mapping work from @jonrohan before we roll this out?

I think this is really good for a PoC. Once that work ships I can integrate it too :)

**Pull request merged**